### PR TITLE
top_k_categorical_accuracy produces wrong result when weighted by sample

### PR DESCRIPTION
**Pull request by MKisilyov**
### Summary

Metrics top_k_categorical_accuracy and sparse_top_k_categorical_accuracy should output a vector of accuracies instead of single scalar to be compatible with weighted_masked_objective function:
https://github.com/keras-team/keras/blob/7c7e51ea5ab47b67cd68374400051dd022bdc662/keras/engine/training_utils.py#L398

weighted_masked_objective function expects to receive a vector with separate accuracy for every sample in batch, it's needed to properly reduce accuracies when sample weight is 0:
https://github.com/keras-team/keras/blob/7c7e51ea5ab47b67cd68374400051dd022bdc662/keras/engine/training_utils.py#L446

If weighted_masked_objective receives a scalar instead of vector, it won't be able to mask samples with weight 0.

Consider this example:
```
y_true = [[0, 1, 0], [1, 0, 0], [0, 0, 1]]
y_pred = [[0, 0.9, 0.1], [0, 0.9, 0.1], [0, 0.9, 0.1]]
sample_weights=[1.0, 0.0, 1.0]

fn = partial(top_k_categorical_accuracy, k=2)

# before fix:
fn(y_true, y_pred) # returns 0.66
weighted_masked_objective(fn)(y_true, y_pred, sample_weights) # returns 0.66

# after fix:
fn(y_true, y_pred) # returns [1.0, 0.0, 1.0]
weighted_masked_objective(fn)(y_true, y_pred, sample_weights) # returns 1.0
```

I'm not sure whether or not this requires a unit test.

### PR Overview

- [?] This PR requires new unit tests [y/n] (make sure tests are included)
- [n] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [y] This PR is backwards compatible [y/n]
- [n] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

**Comment by pavithrasv** [comment]
Thank you for the PR. If you are interested please make the same contribution to tensorflow keras module: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/metrics.py#L2751

**Comment by MKisilyov** [comment]
@pavithrasv 
Thank you for merging this PR!
I tried to do the same fix for TF. But it has different implementation of weighted_masked_objective:
https://github.com/tensorflow/tensorflow/blob/e3f1e41ac6e5dc529b6d40730a0e0298223b3cc8/tensorflow/python/keras/engine/training_utils.py#L673

As far as I understand, it doesn't ignore samples with zero weight. I didn't find any equivalent of `K.not_equal(weights, 0)` in TF version of weighted_masked_objective.
I think, it doesn't make sense to fix top_k accuracy without changing behavior of weighted_masked_objective.

**Comment by pavithrasv** [comment]
@MKisilyov regarding your question about making changes in tensorflow repo, the link you shared is from an old version of tensorflow. The code for weighting is actually here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/utils/losses_utils.py#L163
